en/lessons/basics/collections.md

```diff
@@ -3,209 +3,774 @@ version: 1.3.1
+ghci> adjust (+2) "second" oneItem
+fromList [("first",1)]
+```
 
-```elixir
-iex> map = %{hello: "world"}
-%{hello: "world"}
-iex> %{map | foo: "baz"}
-** (KeyError) key :foo not found in: %{hello: "world"}
-    (stdlib) :maps.update(:foo, "baz", %{hello: "world"})
-    (stdlib) erl_eval.erl:259: anonymous fn/2 in :erl_eval.expr/5
-    (stdlib) lists.erl:1263: :lists.foldl/3
-iex> Map.put(map, :foo, "baz")
-%{foo: "baz", hello: "world"}
+The observant reader will notice that `adjust` doesn't actually update the map.
+The second invocation of adjust returns the original `oneItem` map, this makes
+sense when you consider that all data in haskell is immutable!
+
+## When to use Maps
+
+Maps are really great for in memory persistence of state that will need to be
+retrieved by a key of some arbitrary type. This is is because maps have great
+lookup asymptotics (`O(log n)`) due to the ordering on the key values. The
+example that immediately springs to mind is session storage on the server in a
+web application. The session state can be indexed by an id that is stored in
+the cookie, and the session state can be retrieved from an in memory `Map` on
+every request. This solution doesn't scale infinitely, but you would be
+surprised how well it works!
+
+## Hashmaps
+
+There are some cases where we do not have an ordering on our type, but still
+want to use it as a key to index a map. In this case we probably want to reach
+for a hashmap. A hashmap simply hashes the key, et voilà, we have an ordering!
+This does require that the key type is hashable though.
+
+The module that exports the `HashMap` data type and functionality is called
+`Data.HashMap.Strict`, it lives in the `unordered-containers` package.
+Unfortunately `unordered-containers` are not included in ghci (as of 9.0.1) so
+you will have to setup a project in order to pull in that dependency. Fret not
+though, the api is identical to `Map` aside from a `Hashable k` constraint
+instead of an `Ord k` on the key.
+
+## Vectors
+
+Sometimes we want a list like container where we can have performant indexed
+access. In fact arrays are often the primitive sequential data structure in lots
+of popular programming langauges (rhymes with guava equipped). Well Haskell has
+those kinds of data structures too, and the most popular implementation is the
+`Vector` type from the `vector` library.
+
+One of the coolest things about the `Vector` type is that it offers `O(1)`
+indexed access. The `vector` library offers both a safe and unsafe index
+accessor operator.
+
+Here is an example where we want to create a container that allows us to
+retrieve ascii chars by their character code.
+
+__Note__: Because this library is not a core library it doesn't ship with ghci,
+therefore you will not be able to run the code below. However, if you have
+`cabal-install` available you can run `cabal new-repl --build-depends "vector"`
+and then the snippet below should work.
```

Comment:
Same as above, no need to make a fuss out of it. Just say something along the lines of "you can fetch this package by typing --build-depends when invoking `cabal repl`" (and no need for `new-`, with Cabal 3.4 every command is `v2-` or `new-`)